### ISSUE TITLE
Powerful Chrome Orca - `depositMultipleToken` function does not handle duplicate tokens correctly

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-11-convergence-judging#118]
Powerful Chrome Orca

high

# `depositMultipleToken` function does not handle duplicate tokens correctly

## Summary

Incomplete handling of the `depositMultipleToken` function in `YsDistributor.sol` may cause depositedTokenAddressForTde[_actualTDE] to contain duplicate token addresses.

## Vulnerability Detail

In the function `depositMultipleToken`, argument `deposits` may contain duplicate tokens. However, when `_token` is added to `depositedTokenAddressForTde[_actualTDE]`, `_tokens` (which is initialized with `depositedTokenAddressForTde[_actualTDE]`) is not updated immediately, causing the same token to be added to `depositedTokenAddressForTde[_actualTDE]` multiple times. This violates the uniqueness of the elements in `depositedTokenAddressForTde[_actualTDE]`.

```solidity
function depositMultipleToken(TokenAmount[] calldata deposits) external onlyTreasuryBonds {
        // ....
        address[] memory _tokens = depositedTokenAddressForTde[_actualTDE];
        uint256 tokensLength = _tokens.length;

        for (uint256 i; i < deposits.length; ) {
            IERC20 _token = deposits[i].token;
            uint256 _amount = deposits[i].amount;

            depositedTokenAmountForTde[_actualTDE][_token] += _amount;

            /// @dev Checks whether the token is present in depositedTokenAddressForTde, otherwise we add it.
            bool found;
            for (uint256 j; j < tokensLength; ) {
                if (address(_token) == _tokens[j]) {
                    found = true;
                    break;
                }
                unchecked {
                    ++j;
                }
            }

            if (!found) {
                depositedTokenAddressForTde[_actualTDE].push(address(_token));
            }
	    // ...
        }
        emit DepositedTokens(_actualCycle, deposits);
    }
```

Later, in function `_claimTokenRewards`, since `depositedTokenAddressForTde[tdeId]` may contain duplicate tokens, repeated transfers will result.

```solidity
function _claimTokenRewards(uint256 tokenId, uint256 tdeId, uint256 share, address receiver) internal {
        address[] memory tokens = depositedTokenAddressForTde[tdeId];
        TokenAmount[] memory tokensClaimable = new TokenAmount[](tokens.length);

        for (uint256 i; i < tokens.length; ) {
            IERC20 _token = IERC20(tokens[i]);
            uint256 _amountUser = _calculateUserRewardAmount(tdeId, _token, share);

            tokensClaimable[i] = TokenAmount({token: _token, amount: _amountUser});

            _token.safeTransfer(receiver, _amountUser);
	    // ...
        }
        emit TokensClaim(tokenId, tdeId, share, tokensClaimable);
    }
```

## Impact

Repeated transfers, loss of funds, contract state confusion

## Code Snippet

https://github.com/sherlock-audit/2023-11-convergence/blob/main/sherlock-cvg/contracts/Rewards/YsDistributor.sol#L101

https://github.com/sherlock-audit/2023-11-convergence/blob/main/sherlock-cvg/contracts/Rewards/YsDistributor.sol#L205

## Tool used

Manual Review

## Recommendation

Update the _tokens immediately after updating

```solidity
function depositMultipleToken(TokenAmount[] calldata deposits) external onlyTreasuryBonds {
        // ....
        address[] memory _tokens = depositedTokenAddressForTde[_actualTDE];
        uint256 tokensLength = _tokens.length;

        for (uint256 i; i < deposits.length; ) {
            IERC20 _token = deposits[i].token;
            uint256 _amount = deposits[i].amount;

            depositedTokenAmountForTde[_actualTDE][_token] += _amount;

            /// @dev Checks whether the token is present in depositedTokenAddressForTde, otherwise we add it.
            bool found;
            for (uint256 j; j < tokensLength; ) {
                if (address(_token) == _tokens[j]) {
                    found = true;
                    break;
                }
                unchecked {
                    ++j;
                }
            }

            if (!found) {
                depositedTokenAddressForTde[_actualTDE].push(address(_token));
                _tokens.push(address(_token));  // <----- add
            }
            // ....
        }
    }
```

**Comment by zrax-x** [comment]
Escalate
Maybe I didn't express it clearly, I meant that `tokensLength` was not updated during the check. Therefore, a possible situation is that when the `deposits` contain two same tokens, both will be pushed into `depositedTokenAddressForTde[_actualTDE]`. Of course, if `deposits` cannot contain the same tokens, then this issue is indeed invalid.

**Comment by sherlock-admin2** [comment]
> Escalate
> Maybe I didn't express it clearly, I meant that `tokensLength` was not updated during the check. Therefore, a possible situation is that when the `deposits` contain two same tokens, both will be pushed into `depositedTokenAddressForTde[_actualTDE]`. Of course, if `deposits` cannot contain the same tokens, then this issue is indeed invalid.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by ChechetkinVV** [comment]
There is a duplicate of the issue here: https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209

**Comment by zrax-x** [comment]
Yes, this is a duplicate of [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209). However, after carefully looking at the comment in [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209), it seems that it also misses the point. My doubt is: Is it possible that the uniqueness of the elements in `depositedTokenAddressForTde[_actualTDE]` will be violated?

**Comment by ChechetkinVV** [comment]
> Yes, this is a duplicate of [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209). However, after carefully looking at the comment in [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209), it seems that it also misses the point. My doubt is: Is it possible that the uniqueness of the elements in `depositedTokenAddressForTde[_actualTDE]` will be violated?

I believe that the probability is small, but the impact is significant. Otherwise it would not be worth including in the report.

**Comment by zrax-x** [comment]
> > Yes, this is a duplicate of [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209). However, after carefully looking at the comment in [issue#209](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/209), it seems that it also misses the point. My doubt is: Is it possible that the uniqueness of the elements in `depositedTokenAddressForTde[_actualTDE]` will be violated?
> 
> I believe that the probability is small, but the impact is significant. Otherwise it would not be worth including in the report.

Thanks for your reply. I agree with you. But I still can't understand why the judge marked it as Excluded. Is it because the probability is too small?

**Comment by nevillehuang** [comment]
@zrax-x 

I get what you mean now, but this function can only be called by the trusted admin treasury bonds multisig set by admin [here](https://github.com/sherlock-audit/2023-11-convergence/blob/e894be3e36614a385cf409dc7e278d5b8f16d6f2/sherlock-cvg/contracts/CvgControlTower.sol#L289-L291), so this would constitute admin error that is not valid based on sherlock rules, given admin are trusted to pass in correct inputs while depositing.

**Comment by nevillehuang** [comment]
@zrax-x 

I get what you mean now, but this function can only be called by the trusted admin treasury bonds multisig set by admin [here](https://github.com/sherlock-audit/2023-11-convergence/blob/e894be3e36614a385cf409dc7e278d5b8f16d6f2/sherlock-cvg/contracts/CvgControlTower.sol#L289-L291), so this would constitute admin error that is not valid based on sherlock rules, given admin are trusted to pass in correct inputs while depositing.

@0xR3vert @walk-on-me @shalbe-cvg I believe this is more of a sanity check on already pushed tokens, but a fix could still be made.

**Comment by zrax-x** [comment]
> @zrax-x
> 
> I get what you mean now, but this function can only be called by the trusted admin treasury bonds multisig set by admin [here](https://github.com/sherlock-audit/2023-11-convergence/blob/e894be3e36614a385cf409dc7e278d5b8f16d6f2/sherlock-cvg/contracts/CvgControlTower.sol#L289-L291), so this would constitute admin error that is not valid based on sherlock rules, given admin are trusted to pass in correct inputs while depositing.
> 
> @0xR3vert @walk-on-me @shalbe-cvg I believe this is more of a sanity check on already pushed tokens, but a fix could still be made.

Thank you for your reply. I was confused before because I was not clear about the Sherlock rules, but now I completely understand! I should take a closer look at the Sherlock rules now.

**Comment by Czar102** [comment]
Planning to consider this issue a low and reject the escalation.

**Comment by Czar102** [comment]
Result:
Low
Unique

(duplication status doesn't matter in case of low/invalid issues)

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [zrax-x](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/118/#issuecomment-1868779365): rejected
